File: arcane/installer.py
"""Install merged profile into a target project directory."""

from __future__ import annotations

import json
import shutil
from datetime import datetime, timezone
from pathlib import Path

from .profiles import MergedProfile

SETTINGS_TEMPLATE = {
    "$schema": "https://json.schemastore.org/claude-code-settings.json",
    "statusLine": {
        "type": "command",
        "command": "bash .claude/statusline.sh",
    },
    "permissions": {"allow": [], "deny": []},
    "hooks": {
        "SessionStart": [
            {
                "matcher": "",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/session-start.sh", "timeout": 10},
                    {"type": "command", "command": "bash .claude/hooks/detect-division.sh", "timeout": 10},
                    {"type": "command", "command": "bash .claude/hooks/detect-gaps.sh", "timeout": 10},
                ],
            }
        ],
        "PreToolUse": [
            {
                "matcher": "Bash",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/validate-commit.sh", "timeout": 15},
                    {"type": "command", "command": "bash .claude/hooks/validate-push.sh", "timeout": 10},
                    {"type": "command", "command": "bash .claude/hooks/validate-secrets.sh", "timeout": 10},
                ],
            }
        ],
        "PostToolUse": [
            {
                "matcher": "Write|Edit",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/validate-assets.sh", "timeout": 10},
                    {"type": "command", "command": "bash .claude/hooks/validate-skill-change.sh", "timeout": 5},
                ],
            }
        ],
        "Notification": [
            {
                "matcher": "",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/notify.sh", "timeout": 10},
                ],
            }
        ],
        "PreCompact": [
            {
                "matcher": "",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/pre-compact.sh", "timeout": 10},
                ],
            }
        ],
        "PostCompact": [
            {
                "matcher": "",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/post-compact.sh", "timeout": 10},
                ],
            }
        ],
        "Stop": [
            {
                "matcher": "",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/session-stop.sh", "timeout": 10},
                ],
            }
        ],
        "SubagentStart": [
            {
                "matcher": "",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/log-agent.sh", "timeout": 5},
                ],
            }
        ],
        "SubagentStop": [
            {
                "matcher": "",
                "hooks": [
                    {"type": "command", "command": "bash .claude/hooks/log-agent-stop.sh", "timeout": 5},
                ],
            }
        ],
    },
}


class Installer:
    def __init__(self, arcane_dir: Path, target: Path, merged: MergedProfile, *, dry_run: bool = False):
        self.arcane = arcane_dir
        self.target = target
        self.claude_dir = target / ".claude"
        self.merged = merged
        self.dry_run = dry_run
        self._logs: list[str] = []

    def _log(self, msg: str) -> None:
        self._logs.append(msg)
        print(msg)

    def run(self) -> list[str]:
        self._backup()
        self._create_dirs()
        self._copy_hooks()
        self._copy_skills()
        self._copy_rules()
        self._copy_agents()
        self._copy_docs()
        self._generate_settings()
        self._generate_manifest()
        return self._logs

    def _backup(self) -> None:
        if not self.claude_dir.exists():
            return
        backup = self.target / ".claude.bak"
        if self.dry_run:
            self._log("  [dry-run] Would backup .claude/ -> .claude.bak/")
            return
        if backup.exists():
            shutil.rmtree(backup)
        shutil.move(str(self.claude_dir), str(backup))
        self._log("  Backed up .claude/ -> .claude.bak/")

    def _create_dirs(self) -> None:
        if self.dry_run:
            return
        for d in ["skills", "rules", "agents", "docs"]:
            (self.claude_dir / d).mkdir(parents=True, exist_ok=True)

    def _copy_hooks(self) -> None:
        self._log("\nHooks:")
        hooks_src = self.arcane / ".claude" / "hooks"
        statusline_src = self.arcane / ".claude" / "statusline.sh"

        if self.dry_run:
            self._log("  [dry-run] hooks/ -> .claude/hooks/")
            self._log("  [dry-run] statusline.sh -> .claude/statusline.sh")
            return

        if hooks_src.is_dir():
            shutil.copytree(str(hooks_src), str(self.claude_dir / "hooks"), dirs_exist_ok=True)
            self._log("  [ok] hooks/")

        if statusline_src.is_file():
            shutil.copy2(str(statusline_src), str(self.claude_dir / "statusline.sh"))
            self._log("  [ok] statusline.sh")

    def _find_skill_source(self, skill: str) -> Path | None:
        """Auto-discover skill across all skills-* pool directories."""
        for pool_dir in sorted((self.arcane / ".claude").iterdir()):
            if pool_dir.is_dir() and pool_dir.name.startswith("skills-"):
                candidate = pool_dir / skill
                if candidate.is_dir():
                    return candidate
        return None

    def _copy_skills(self) -> None:
        self._log("\nSkills:")
        for skill in self.merged.skills:
            src = self._find_skill_source(skill)
            if src is None:
                self._log(f"  WARN: Skill '{skill}' not found in any skills-* pool")
                continue
            dst = self.claude_dir / "skills" / skill
            if self.dry_run:
                self._log(f"  [dry-run] {src.parent.name}/{skill} -> skills/{skill}")
            else:
                shutil.copytree(str(src), str(dst), dirs_exist_ok=True)
                self._log(f"  [ok] {skill}")

        if self.merged.has_gamedev:
            templates_src = self.arcane / ".claude" / "skills-gamedev" / "_templates"
            if templates_src.is_dir():
                if self.dry_run:
                    self._log("  [dry-run] _templates/")
                else:
                    dst = self.claude_dir / "skills" / "_templates"
                    shutil.copytree(str(templates_src), str(dst), dirs_exist_ok=True)
                    self._log("  [ok] _templates/")

    def _copy_rules(self) -> None:
        self._log("\nRules:")
        for rule in self.merged.rules_universal:
            src = self.arcane / ".claude" / "rules" / f"{rule}.md"
            dst = self.claude_dir / "rules" / f"{rule}.md"
            if not src.is_file():
                self._log(f"  WARN: Rule '{rule}.md' not found")
                continue
            if self.dry_run:
                self._log(f"  [dry-run] {rule}.md")
            else:
                shutil.copy2(str(src), str(dst))
                self._log(f"  [ok] {rule}.md")

        for rule in self.merged.rules_gamedev:
            src = self.arcane / ".claude" / "skills-gamedev" / "_rules" / f"{rule}.md"
            dst = self.claude_dir / "rules" / f"{rule}.md"
            if not src.is_file():
                self._log(f"  WARN: Rule '{rule}.md' not found")
                continue
            if self.dry_run:
                self._log(f"  [dry-run] {rule}.md")
            else:
                shutil.copy2(str(src), str(dst))
                self._log(f"  [ok] {rule}.md")

    def _copy_agents(self) -> None:
        if not self.merged.agents:
            return
        self._log("\nAgents:")
        for agent_dir in self.merged.agents:
            src = self.arcane / ".claude" / "agents" / agent_dir
            if not src.is_dir():
                src = self.arcane / "agents" / agent_dir
            if not src.is_dir():
                self._log(f"  WARN: Agents dir '{agent_dir}' not found")
                continue
            dst = self.claude_dir / "agents" / agent_dir
            if self.dry_run:
                count = sum(1 for _ in src.rglob("*.md"))
                self._log(f"  [dry-run] agents/{agent_dir}/ ({count} agents)")
            else:
                shutil.copytree(str(src), str(dst), dirs_exist_ok=True)
                count = sum(1 for _ in dst.rglob("*.md"))
                self._log(f"  [ok] agents/{agent_dir}/ ({count} agents)")

    def _copy_docs(self) -> None:
        docs_src = self.arcane / ".claude" / "docs"
        if not docs_src.is_dir():
            return
        if self.dry_run:
            self._log("\nDocs:")
            self._log("  [dry-run] docs/")
            return
        dst = self.claude_dir / "docs"
        shutil.copytree(str(docs_src), str(dst), dirs_exist_ok=True)
        self._log("\nDocs:")
        self._log("  [ok] docs/")

    def _generate_settings(self) -> None:
        self._log("\nConfig:")
        settings = json.loads(json.dumps(SETTINGS_TEMPLATE))
        settings["permissions"]["allow"] = list(self.merged.permissions_allow)
        settings["permissions"]["deny"] = list(self.merged.permissions_deny)

        if self.dry_run:
            self._log(
                f"  [dry-run] settings.json ({len(self.merged.permissions_allow)} allow, "
                f"{len(self.merged.permissions_deny)} deny)"
            )
            return

        path = self.claude_dir / "settings.json"
        path.write_text(json.dumps(settings, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
        self._log("  [ok] settings.json")

    def _generate_manifest(self) -> None:
        profile_cmd = "+".join(p for p in self.merged.loaded if p != "core")
        manifest = {
            "arcane_version": "1.0.0",
            "profile_command": profile_cmd,
            "profiles": self.merged.loaded,
            "installed_at": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
            "total_skills": self.merged.total_skills,
            "total_rules": self.merged.total_rules,
            "source": str(self.arcane),
        }

        if self.dry_run:
            self._log("  [dry-run] arcane-manifest.json")
            return

        path = self.claude_dir / "arcane-manifest.json"
        path.write_text(json.dumps(manifest, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
        self._log("  [ok] arcane-manifest.json")


def clean(target: Path) -> None:
    claude_dir = target / ".claude"
    backup = target / ".claude.bak"
    if not claude_dir.exists():
        print(f"No .claude/ found in {target} — nothing to clean.")
        return

    manifest = claude_dir / "arcane-manifest.json"
    if manifest.exists():
        data = json.loads(manifest.read_text(encoding="utf-8"))
        profiles = [p for p in data.get("profiles", []) if p != "core"]
        print(f"  Current profile: {' + '.join(profiles)}")
        print(f"  Installed: {data.get('installed_at', 'unknown')}")

    shutil.rmtree(str(claude_dir))
    if backup.exists():
        shutil.rmtree(str(backup))
    print(f"\n  [ok] Removed .claude/ from {target}")

File: src/installer.ts
import fs from "node:fs";
import path from "node:path";
import type { MergedProfile, InstallerOptions } from "./types.js";
import {
  copyDirSync,
  ensureDir,
  getPackageRoot,
  isSymlinkOrJunction,
  safeRemove,
} from "./utils.js";
import { writeManifest } from "./manifest.js";
import { linkOrCopyDir } from "./worktree.js";
import { computeContentHashes } from "./content-hash.js";

const SETTINGS_TEMPLATE = {
  $schema: "https://json.schemastore.org/claude-code-settings.json",
  permissions: { allow: [] as string[], deny: [] as string[] },
  hooks: {
    SessionStart: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/session-start.sh",
            timeout: 10,
          },
          {
            type: "command",
            command: "bash .claude/hooks/detect-division.sh",
            timeout: 10,
          },
          {
            type: "command",
            command: "bash .claude/hooks/detect-gaps.sh",
            timeout: 10,
          },
          {
            type: "command",
            command: "bash .claude/hooks/check-update.sh",
            timeout: 10,
          },
        ],
      },
    ],
    PreToolUse: [
      {
        matcher: "Bash",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/validate-commit.sh",
            timeout: 15,
          },
          {
            type: "command",
            command: "bash .claude/hooks/validate-push.sh",
            timeout: 10,
          },
          {
            type: "command",
            command: "bash .claude/hooks/validate-secrets.sh",
            timeout: 10,
          },
        ],
      },
    ],
    PostToolUse: [
      {
        matcher: "Write|Edit",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/validate-assets.sh",
            timeout: 10,
          },
          {
            type: "command",
            command: "bash .claude/hooks/validate-skill-change.sh",
            timeout: 5,
          },
        ],
      },
    ],
    Notification: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/notify.sh",
            timeout: 10,
          },
        ],
      },
    ],
    PreCompact: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/pre-compact.sh",
            timeout: 10,
          },
        ],
      },
    ],
    PostCompact: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/post-compact.sh",
            timeout: 10,
          },
        ],
      },
    ],
    Stop: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/session-stop.sh",
            timeout: 10,
          },
        ],
      },
    ],
    SubagentStart: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/log-agent.sh",
            timeout: 5,
          },
        ],
      },
    ],
    SubagentStop: [
      {
        matcher: "",
        hooks: [
          {
            type: "command",
            command: "bash .claude/hooks/log-agent-stop.sh",
            timeout: 5,
          },
        ],
      },
    ],
  },
};

export class Installer {
  private root: string;
  private target: string;
  private claudeDir: string;
  private merged: MergedProfile;
  private dryRun: boolean;
  private logs: string[] = [];

  private shareFrom?: string;
  private sharedDirs: string[] = [];

  constructor(merged: MergedProfile, opts: InstallerOptions) {
    this.root = opts.contentRoot ?? getPackageRoot();
    this.target = opts.target;
    this.claudeDir = path.join(opts.target, ".claude");
    this.merged = merged;
    this.dryRun = opts.dryRun;
    this.shareFrom = opts.shareFrom;
  }

  private log(msg: string): void {
    this.logs.push(msg);
    console.log(msg);
  }

  run(
    profileCommand: string,
    worktreeMeta?: {
      is_worktree: boolean;
      main_worktree: string;
    },
  ): string[] {
    this.backup();
    this.createDirs();
    this.copyHooks();
    this.copySkills();
    this.copyRules();
    this.copyAgents();
    this.copyDocs();
    this.generateSettings();

    if (!this.dryRun) {
      const wt =
        worktreeMeta && this.sharedDirs.length > 0
          ? { ...worktreeMeta, shared_dirs: this.sharedDirs }
          : worktreeMeta
            ? { ...worktreeMeta, shared_dirs: [] }
            : undefined;
      const contentHashes = computeContentHashes(this.claudeDir);
      writeManifest(this.target, this.merged, profileCommand, this.root, {
        worktree: wt,
        contentHashes,
      });
      this.log("  [ok] arcane-manifest.json");
    } else {
      this.log("  [dry-run] arcane-manifest.json");
    }

    return this.logs;
  }

  private backup(): void {
    if (!fs.existsSync(this.claudeDir)) return;
    const backup = path.join(this.target, ".claude.bak");
    if (this.dryRun) {
      this.log("  [dry-run] Would backup .claude/ -> .claude.bak/");
      return;
    }
    if (fs.existsSync(backup)) {
      fs.rmSync(backup, { recursive: true, force: true });
    }
    fs.renameSync(this.claudeDir, backup);
    this.log("  Backed up .claude/ -> .claude.bak/");
  }

  private createDirs(): void {
    if (this.dryRun) return;
    for (const d of ["skills", "rules", "agents", "docs"]) {
      ensureDir(path.join(this.claudeDir, d));
    }
  }

  private get hasStatusline(): boolean {
    return this.merged.loaded.includes("statusline");
  }

  private copyHooks(): void {
    this.log("\nHooks:");
    const hooksDir = path.join(this.claudeDir, "hooks");

    if (this.dryRun) {
      if (this.shareFrom) {
        this.log("  [dry-run] hooks/ -> symlink from main worktree");
      } else {
        this.log("  [dry-run] hooks/ -> .claude/hooks/");
      }
      if (this.hasStatusline) this.log("  [dry-run] statusline.sh");
      return;
    }

    if (this.shareFrom) {
      const sharedHooks = path.join(this.shareFrom, ".claude", "hooks");
      if (fs.existsSync(sharedHooks)) {
        const result = linkOrCopyDir(sharedHooks, hooksDir, "symlink");
        if (result === "linked") {
          this.sharedDirs.push("hooks");
          this.log("  [ok] hooks/ (shared via symlink)");
        } else {
          this.log("  [ok] hooks/ (copied — symlink failed)");
        }
        this.copyStatusline();
        return;
      }
    }

    const hooksSrc = path.join(this.root, "hooks");
    if (fs.existsSync(hooksSrc)) {
      ensureDir(hooksDir);
      for (const entry of fs.readdirSync(hooksSrc, { withFileTypes: true })) {
        if (entry.name === "statusline.sh") continue;
        const src = path.join(hooksSrc, entry.name);
        const dst = path.join(hooksDir, entry.name);
        if (entry.isDirectory()) {
          copyDirSync(src, dst);
        } else {
          fs.copyFileSync(src, dst);
        }
      }
      this.log("  [ok] hooks/");
    }
    this.copyStatusline();
  }

  private copyStatusline(): void {
    if (!this.hasStatusline) return;
    const statusline = path.join(this.root, "hooks", "statusline.sh");
    if (fs.existsSync(statusline)) {
      fs.copyFileSync(
        statusline,
        path.join(this.claudeDir, "statusline.sh"),
      );
      this.log("  [ok] statusline.sh (+statusline addon)");
    }
  }

  private copySkills(): void {
    this.log("\nSkills:");
    for (const skill of this.merged.skills) {
      const src = path.join(this.root, "skills", skill);
      if (!fs.existsSync(src)) {
        this.log(`  WARN: Skill '${skill}' not found in skills/`);
        continue;
      }
      const dst = path.join(this.claudeDir, "skills", skill);
      if (this.dryRun) {
        this.log(`  [dry-run] ${skill}`);
      } else {
        copyDirSync(src, dst);
        this.log(`  [ok] ${skill}`);
      }
    }

    const hasGamedev = this.merged.rules.gamedev.length > 0;
    if (hasGamedev) {
      const tplSrc = path.join(this.root, "templates", "gamedev");
      if (fs.existsSync(tplSrc)) {
        const tplDst = path.join(this.claudeDir, "skills", "_templates");
        if (this.dryRun) {
          this.log("  [dry-run] _templates/");
        } else {
          copyDirSync(tplSrc, tplDst);
          this.log("  [ok] _templates/");
        }
      }
    }
  }

  private copyRules(): void {
    this.log("\nRules:");
    for (const rule of this.merged.rules.universal) {
      const src = path.join(this.root, "rules", `${rule}.md`);
      if (!fs.existsSync(src)) {
        this.log(`  WARN: Rule '${rule}.md' not found`);
        continue;
      }
      if (this.dryRun) {
        this.log(`  [dry-run] ${rule}.md`);
      } else {
        fs.copyFileSync(
          src,
          path.join(this.claudeDir, "rules", `${rule}.md`),
        );
        this.log(`  [ok] ${rule}.md`);
      }
    }
    for (const rule of this.merged.rules.gamedev) {
      const src = path.join(this.root, "rules", "gamedev", `${rule}.md`);
      if (!fs.existsSync(src)) {
        this.log(`  WARN: Rule '${rule}.md' not found`);
        continue;
      }
      if (this.dryRun) {
        this.log(`  [dry-run] ${rule}.md`);
      } else {
        fs.copyFileSync(
          src,
          path.join(this.claudeDir, "rules", `${rule}.md`),
        );
        this.log(`  [ok] ${rule}.md`);
      }
    }
  }

  private copyAgents(): void {
    if (this.merged.agents.length === 0) return;
    this.log("\nAgents:");
    for (const agentDir of this.merged.agents) {
      const src = path.join(this.root, "agents", agentDir);
      if (!fs.existsSync(src)) {
        this.log(`  WARN: Agent dir '${agentDir}' not found`);
        continue;
      }
      const dst = path.join(this.claudeDir, "agents", agentDir);
      if (this.dryRun) {
        this.log(`  [dry-run] agents/${agentDir}/`);
      } else {
        copyDirSync(src, dst);
        const count = fs
          .readdirSync(dst, { recursive: true })
          .filter((f) => String(f).endsWith(".md")).length;
        this.log(`  [ok] agents/${agentDir}/ (${count} agents)`);
      }
    }
  }

  private copyDocs(): void {
    const docsDir = path.join(this.claudeDir, "docs");
    const docsSrc = path.join(this.root, "docs");
    if (!fs.existsSync(docsSrc)) return;
    this.log("\nDocs:");
    if (this.dryRun) {
      if (this.shareFrom) {
        this.log("  [dry-run] docs/ -> symlink from main worktree");
      } else {
        this.log("  [dry-run] docs/");
      }
      return;
    }

    if (this.shareFrom) {
      const sharedDocs = path.join(this.shareFrom, ".claude", "docs");
      if (fs.existsSync(sharedDocs)) {
        const result = linkOrCopyDir(sharedDocs, docsDir, "symlink");
        if (result === "linked") {
          this.sharedDirs.push("docs");
          this.log("  [ok] docs/ (shared via symlink)");
        } else {
          this.log("  [ok] docs/ (copied — symlink failed)");
        }
        return;
      }
    }

    copyDirSync(docsSrc, docsDir);
    this.log("  [ok] docs/");
  }

  private generateSettings(): void {
    this.log("\nConfig:");
    const settings = JSON.parse(JSON.stringify(SETTINGS_TEMPLATE));
    settings.permissions.allow = [...this.merged.permissions.allow];
    settings.permissions.deny = [...this.merged.permissions.deny];
    if (this.hasStatusline) {
      settings.statusLine = {
        type: "command",
        command: "bash .claude/statusline.sh",
      };
    }

    if (this.dryRun) {
      this.log(
        `  [dry-run] settings.json (${this.merged.permissions.allow.length} allow, ${this.merged.permissions.deny.length} deny)`,
      );
      return;
    }
    const settingsPath = path.join(this.claudeDir, "settings.json");
    fs.writeFileSync(
      settingsPath,
      JSON.stringify(settings, null, 2) + "\n",
      "utf-8",
    );
    this.log("  [ok] settings.json");
  }
}

export function clean(target: string): void {
  const claudeDir = path.join(target, ".claude");
  const backup = path.join(target, ".claude.bak");
  if (!fs.existsSync(claudeDir)) {
    console.log(`No .claude/ found in ${target} — nothing to clean.`);
    return;
  }

  const manifestFile = path.join(claudeDir, "arcane-manifest.json");
  if (fs.existsSync(manifestFile)) {
    const data = JSON.parse(fs.readFileSync(manifestFile, "utf-8"));
    const profiles = (data.profiles ?? []).filter(
      (p: string) => p !== "core",
    );
    console.log(`  Current profile: ${profiles.join(" + ")}`);
    console.log(`  Installed: ${data.installed_at ?? "unknown"}`);

    const shared = data.worktree?.shared_dirs ?? [];
    for (const dir of shared) {
      const p = path.join(claudeDir, dir);
      if (isSymlinkOrJunction(p)) {
        fs.unlinkSync(p);
        console.log(`  [ok] Unlinked shared ${dir}/`);
      }
    }
  }

  for (const sub of ["hooks", "docs", "skills", "agents", "rules"]) {
    const p = path.join(claudeDir, sub);
    safeRemove(p);
  }

  const remaining = fs.readdirSync(claudeDir);
  for (const f of remaining) {
    const p = path.join(claudeDir, f);
    safeRemove(p);
  }

  if (fs.existsSync(claudeDir)) {
    fs.rmSync(claudeDir, { recursive: true, force: true });
  }
  if (fs.existsSync(backup)) {
    fs.rmSync(backup, { recursive: true, force: true });
  }
  console.log(`\n  [ok] Removed .claude/ from ${target}`);
}
